### ISSUE TITLE
Compiler panic on type inference for impl Generator yielding a reference

### ISSUE THREAD

**Issue by bergus** [rust-lang/rust#52304]
I'm getting

> error: internal compiler error: librustc/infer/error_reporting/mod.rs:184: ReEmpty
> thread 'main' panicked at 'Box<Any>', librustc_errors/lib.rs:554:9

when trying to compile
```rust
#![feature(generators, generator_trait)]
use std::ops::Generator;
fn example() -> impl Generator {
    || { yield &1 }
}
```
on rustc 1.29.0-nightly (e5f6498d3 2018-07-10) - [Playground demo](https://play.rust-lang.org/?gist=2421dbda0b8b259b3efec9897a58f2d6&version=nightly&mode=debug&edition=2015).

This seems to happen when I `yield` a reference type and don't specify the exact type in `impl Generator<Yield=&Whatever>`.

**Comment by Zoxc** [comment]
We don't generate an error message for the `ReEmpty` region here and panic instead. Here is the message that should be printed with a placeholder:
```
error[E0700]: hidden type for `impl Trait` captures lifetime that does not appear in bounds
 --> test-region.rs:3:17
  |
3 | fn example() -> impl Generator {
  |                 ^^^^^^^^^^^^^^
  |
  = note: hidden type `[generator@test-region.rs:4:5: 4:20 for<'r> {i32, &'r i32, ()}]` captures BUG[ReEmpty]
```
Should `ReEmpty` be allowed for impl Trait? cc @rust-lang/compiler

**Comment by arielb1** [comment]
Why are we capturing `ReEmpty` here?

**Comment by Zoxc** [comment]
There isn't any `ReEmpty` in the type. There's some [scary code](https://github.com/rust-lang/rust/blob/master/src/librustc/infer/anon_types/mod.rs#L595) which replaces regions with `ReEmpty` which is closure related. I'm not sure what is going on there, but perhaps generators should be treated the same? cc @nikomatsakis

**Comment by eddyb** [comment]
@Zoxc Oh, all of that code is for exporting `impl Trait` concrete types from inference results, which needs to "connect up" lifetimes inside the inferred concrete type with lifetime parameters.

**Comment by arielb1** [comment]
I think the problem is that the `yield_ty` of the generator ends up being `&'ReEmpty u32`, because nothing is constraining it to be anything else. I wonder why this isn't occurring with closures.

**Comment by arielb1** [comment]
This doesn't occur with closures because regionck forces the return type of the closure to outlive the closure's call scope, which prevents `ReEmpty` from being inferred.

**Comment by arielb1** [comment]
I *think* that the correct way to fix it would be to require all projections in `impl Trait` types to be well-formed.

**Comment by arielb1** [comment]
The requirement we want is that that "implicit" `impl Generator for [generator type]` makes sense.

That requires that the projections in that impl are alive as long as `[generator type]` is alive (we don't consider that as a "well-formed" requirement, but the "projection outlives" rules assume that). That does not hold if regions in the projections can be inferred to `ReEmpty` - we need to be checking that.

**Comment by arielb1** [comment]
I think that in closures, `CallSite` (from the fix to #29793) is serving that rule with lexical lifetimes - not sure what is checking that with NLL.

**Comment by arielb1** [comment]
This can, in fact, be used as a soundness hole (assuming you can call `resume`):
```Rust
#![feature(generators, generator_trait)]
use std::fmt::Debug;

use std::ops::Generator;
fn gen_taker<G: Generator>(mut g: G)
    where G::Yield: Debug, G::Return: Debug
{
    let a = unsafe { g.resume() };
    drop(g);
    println!("{:?}", a);
}
fn main() {
    gen_taker(|| {
        let x = 1;
        yield &x
    });
}
```

**Comment by arielb1** [comment]
Note: the version above is fixed by NLL. need to check whether fix is tight.

**Comment by rossmacarthur** [comment]
@arielb1 any ideas on how to check whether the "fix is tight"? Otherwise perhaps this issue can be closed?
